### PR TITLE
Take `CUDA_VISIBLE_DEVICES` into account for nvml calls

### LINKED ISSUE #94472: `torch.cuda.utilization()` does not work when `CUDA_VISIBLE_DEVICES` is set
### 🐛 Describe the bug

When monitoring from a different process torch.cuda.utilization() returns zero unless `CUDA_VISIBLE_DEVICES` unset or set to `0`.

To reproduce:

monitor.py
```py
import torch
import time

while True:
    print(torch.cuda.utilization())
    time.sleep(1)
```

work.py
```py
import torch

while True:
    t = torch.ones(10000, 10000).cuda()
    t = t * t
```

Results:

```
$ export CUDA_VISIBLE_DEVICES=0; python monitor.py & python work.py & sleep 10; kill %1 %2
0
9
17
12
14
18
17
16
14
```

```
$ export CUDA_VISIBLE_DEVICES=1; python monitor.py & python work.py & sleep 10; kill %1 %2
0
0
0
0
0
0
0
0
0
```

### Versions

Collecting environment information...
PyTorch version: 1.13.1+cu117
Is debug build: False
CUDA used to build PyTorch: 11.7
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.5 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0
Clang version: Could not collect
CMake version: version 3.16.3
Libc version: glibc-2.31

Python version: 3.10.9 (main, Jan 11 2023, 15:21:40) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-5.15.0-56-generic-x86_64-with-glibc2.31
Is CUDA available: True
CUDA runtime version: 11.8.89
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration:
GPU 0: NVIDIA A100-SXM4-40GB
GPU 1: NVIDIA A100-SXM4-40GB
GPU 2: NVIDIA A100-SXM4-40GB
GPU 3: NVIDIA A100-SXM4-40GB
GPU 4: NVIDIA A100-SXM4-40GB
GPU 5: NVIDIA A100-SXM4-40GB
GPU 6: NVIDIA A100-SXM4-40GB
GPU 7: NVIDIA A100-SXM4-40GB

Nvidia driver version: 520.61.05
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Byte Order:                      Little Endian
Address sizes:                   48 bits physical, 48 bits virtual
CPU(s):                          124
On-line CPU(s) list:             0-123
Thread(s) per core:              1
Core(s) per socket:              1
Socket(s):                       124
NUMA node(s):                    2
Vendor ID:                       AuthenticAMD
CPU family:                      23
Model:                           49
Model name:                      AMD EPYC 7542 32-Core Processor
Stepping:                        0
CPU MHz:                         2899.998
BogoMIPS:                        5799.99
Virtualization:                  AMD-V
Hypervisor vendor:               KVM
Virtualization type:             full
L1d cache:                       7.8 MiB
L1i cache:                       7.8 MiB
L2 cache:                        62 MiB
L3 cache:                        2 GiB
NUMA node0 CPU(s):               0-31,64-95
NUMA node1 CPU(s):               32-63,96-123
Vulnerability Itlb multihit:     Not affected
Vulnerability L1tf:              Not affected
Vulnerability Mds:               Not affected
Vulnerability Meltdown:          Not affected
Vulnerability Mmio stale data:   Not affected
Vulnerability Retbleed:          Mitigation; untrained return thunk; SMT disabled
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:        Mitigation; Retpolines, IBPB conditional, STIBP disabled, RSB filling, PBRSB-eIBRS Not affected
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Not affected
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm rep_good nopl cpuid extd_apicid tsc_known_freq pni pclmulqdq ssse3 fma cx16 sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand hypervisor lahf_lm cmp_legacy svm cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw perfctr_core ssbd ibrs ibpb stibp vmmcall fsgsbase tsc_adjust bmi1 avx2 smep bmi2 rdseed adx smap clflushopt clwb sha_ni xsaveopt xsavec xgetbv1 xsaves clzero xsaveerptr wbnoinvd arat npt nrip_save umip rdpid arch_capabilities

Versions of relevant libraries:
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.23.5
[pip3] rotary-embedding-torch==0.1.5
[pip3] torch==1.13.1+cu117
[pip3] torchaudio==0.13.1+cu117
[pip3] torchvision==0.14.1+cu117
[conda] numpy                     1.23.5                   pypi_0    pypi
[conda] rotary-embedding-torch    0.1.5                    pypi_0    pypi
[conda] torch                     1.13.1+cu117             pypi_0    pypi
[conda] torchaudio                0.13.1+cu117             pypi_0    pypi
[conda] torchvision               0.14.1+cu117             pypi_0    pypi

cc @ngimel

### PR DESCRIPTION
**Pull request by jnordberg**
Fixes #94472

cc @ngimel

### CODE REVIEW

File: torch/cuda/__init__.py [hunk 1]
```diff
@@ -544,6 +544,15 @@ def _device_count_nvml() -> int:
         return raw_cnt
     return len(set(range(raw_cnt)).intersection(visible_devices))
 
+def _get_nvml_device_index(device: Optional[Union[int, Device]]) -> int:
+    r"""Returns the NVML index of the device, taking CUDA_VISIBLE_DEVICES into account."""
+    idx = _get_device_index(device, optional=True)
+    visible_devices = _parse_visible_devices()
+    idx_map = {idx: real_idx for idx, real_idx in enumerate(sorted(visible_devices))}
```
Comment (XuehaiPan) [inline torch/cuda/__init__.py:551]: The `visible_devices` should not be sorted.

Comment (jnordberg) [reply torch/cuda/__init__.py:551]: It's a `set`, it has to be sorted or the order is arbitrary

Comment (XuehaiPan) [reply torch/cuda/__init__.py:551]: It shouldn't be a `set`. The implementation of `_parse_visible_devices` is wrong. `CUDA_VISIBLE_DEVICES='0,1'` and `CUDA_VISIBLE_DEVICES='1,0'` have different meanings.

Comment (jnordberg) [reply torch/cuda/__init__.py:551]: Ok, I assumed that was normalized in by the C++ code or CUDA libraries. I can refactor that function to return a list with the order preserved if that's not the case

File: torch/cuda/__init__.py [hunk 2]
```diff
@@ -506,10 +506,13 @@ def _strtoul(s: str) -> int:
...
-    rc: Set[int] = set()
+    rv: List[int] = []
     for elem in var.split(","):
-        rc.add(_strtoul(elem.strip()))
-    return rc
+        idx = _strtoul(elem.strip())
+        if idx in rv:
+            continue
+        rv.append(idx)
+    return rv
```
Comment (jnordberg) [inline torch/cuda/__init__.py:515]: @XuehaiPan I've confirmed that this works also with settings like `CUDA_VISIBLE_DEVICES=1,0`

File: torch/cuda/__init__.py [hunk 3]
```diff
@@ -531,7 +534,7 @@ def _raw_device_count_nvml() -> int:
 def _device_count_nvml() -> int:
     """Return number of devices as reported by NVML taking CUDA_VISIBLE_DEVICES into account.
     Negative value is returned if NVML discovery or initialization has failed."""
-    visible_devices = _parse_visible_devices()
+    visible_devices = set(_parse_visible_devices())
```
Comment (XuehaiPan) [inline torch/cuda/__init__.py:537]:
```suggestion
    visible_devices = _parse_visible_devices()
```

`set.intersection(self, Iterable)` accepts `list`.

File: torch/cuda/__init__.py [hunk 4]
```diff
@@ -487,11 +487,11 @@ def set_stream(stream: Stream):
         return
     torch._C._cuda_setStream(stream_id=stream.stream_id, device_index=stream.device_index, device_type=stream.device_type)
 
-def _parse_visible_devices() -> Set[int]:
+def _parse_visible_devices() -> List[int]:
     """Parse CUDA_VISIBLE_DEVICES environment variable."""
     var = os.getenv("CUDA_VISIBLE_DEVICES")
     if var is None:
-        return set(x for x in range(64))
+        return list(x for x in range(64))
```
Comment (XuehaiPan) [inline torch/cuda/__init__.py:494]:
```suggestion
        return list(range(64))
```

File: torch/cuda/__init__.py (hunk 2 above)
Comment (XuehaiPan) [reply torch/cuda/__init__.py:515]: I think it's fine to apply this change in this PR. Further bug fixes may be done in #92315. cc @malfet

- #92315

**Comment by ngimel** [comment]
Ugh, I think we should wait for #92315 to get merged to have a better way of parsing visible devices. 
It's true that visible devices shouldn't be a set because order matters, but there are also annoying quirks in CUDA_VISIBLE_DEVICES (e.g. if some number is repeated, it results in no visible devices).

**Comment by jnordberg** [comment]
@ngimel The behavior should be the same as before with regards to duplicate devices, I copied the behavior of a set by not including items in the list that are already there

**Comment by ngimel** [comment]
Right, but #92315 fixes a few bugs in `_parse_visible_devices`, so it'll be good to get it in first, I've pinged the author

File: torch/cuda/__init__.py [hunk 5]
```diff
@@ -544,6 +547,15 @@ def _device_count_nvml() -> int:
         return raw_cnt
     return len(set(range(raw_cnt)).intersection(visible_devices))
 
+def _get_nvml_device_index(device: Optional[Union[int, Device]]) -> int:
+    r"""Returns the NVML index of the device, taking CUDA_VISIBLE_DEVICES into account."""
+    idx = _get_device_index(device, optional=True)
+    visible_devices = _parse_visible_devices()
```
Comment (malfet) [inline torch/cuda/__init__.py:553]: Q: Should this also work for MIG- prefixed devices, as described in https://github.com/pytorch/pytorch/issues/90543 ?

Comment (XuehaiPan) [reply torch/cuda/__init__.py:553]: This is not working with MIG devices. The device handle is retrieved by NVML index:

```python
handle = pynvml.nvmlDeviceGetHandleByIndex(device)
```

If we want to support MIG devices here, it's better to use UUIDs:

```python
uuid = _get_nvml_device_uuid(device)
handle = pynvml.nvmlDeviceGetHandleByUUID(uuid)
```

then the returned can be a GPU device handle or a MIG device handle.

**Comment by ngimel** [comment]
#92315 is landed, @jnordberg can you please rebase?

**Review by ngimel — APPROVED** [review]
Awesome, thanks!

**Pull request closed without merging**